### PR TITLE
Fix loading tk on windows when current process has >1024 modules.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

Closes https://github.com/matplotlib/matplotlib/issues/22378.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by QuLogic** [comment]
Were you able to test this out? One of the key questions from the issue I had is what the API actually does for too-small buffers, as it was unclear from the docs whether it failed, truncated, or something in between.

It looks like tests are okay, but the problem with memory corruption is you can never be sure it's not accidentally working.

**Comment by anntzer** [comment]
Sorry, I could not exercise the pathological case of too-many-modules.  OTOH, even though the behavior for too small arrays is not explicitly stated, I think it is implicit in the docs (https://docs.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-enumprocessmodules):
> To determine if the lphModule array is too small to hold all module handles for the process, compare the value returned in lpcbNeeded with the value specified in cb. If lpcbNeeded is greater than cb, increase the size of the array and call EnumProcessModules again.

This strongly suggests that it is OK to call EnumProcessModules with a too-small array.

**Comment by QuLogic** [comment]
I guess the question is more whether it returns an error or not, which would avoid the potential read out of bounds in the original issue, but then also exit early from what you've written. Or maybe it's clever and returns no error with `NULL` buffer.

**Review by tacaswell — APPROVED** [review]

**Comment by anntzer** [comment]
> Or maybe it's clever and returns no error with NULL buffer.

I did try that this works to load tk on a standalone process, which means that the first call to EnumProcessModules does not error out.

**Comment by javidcf** [comment]
> @javidcf can you see if this fixes the problem for you?

Thanks, looks ok as far as I can tell, although I don't have the setup to test my own build of Matplotlib, and unfortunately I probably won't be able to set it up soon as I am currently on leave.

If the context helps, this happened when I was trying to use Matplotlib within the Python interpreter in Unreal Engine, with a lot of engine plugins loaded.

**Comment by tacaswell** [comment]
> ...when I was trying to use Matplotlib within the Python interpreter in Unreal Engine....

That is a sentence I never expected to read.  Are you willing to share any more context?  Also in that context do you really need tk windows?

**Comment by javidcf** [comment]
> > ...when I was trying to use Matplotlib within the Python interpreter in Unreal Engine....
> 
> That is a sentence I never expected to read. Are you willing to share any more context? Also in that context do you really need tk windows?

I don't need Tk windows in particular, I was trying to use Matplotlib to do some visualization and manipulation of data coming from Unreal Engine (e.g. from assets and other things, for tooling purposes). It works, but the default TkAgg backend fails to load, so at the moment I use Qt5Agg.

File: src/_tkagg.cpp [hunk 1]
```diff
@@ -256,25 +256,32 @@ int load_tcl(T lib)
...
-    DWORD cbNeeded;
-    unsigned int i;
+    HANDLE process = GetCurrentProcess();  // Pseudo-handle, doesn't need closing.
+    HMODULE* modules = NULL;
+    DWORD size;
+    if (!EnumProcessModules(process, NULL, 0, &size)) {
+        PyErr_SetFromWindowsErr(0);
+        goto exit;
+    }
+    if (!(modules = reinterpret_cast<HMODULE*>(malloc(size)))) {
```
Comment (QuLogic) [inline src/_tkagg.cpp:266]: It's more of a `static_cast`, isn't it? `void*` has no form and thus can't be (re)interpreted.

**Review by QuLogic — APPROVED** [review]

File: src/_tkagg.cpp (hunk 1 above)
Comment (anntzer) [reply src/_tkagg.cpp]: Ah, yes.

**Comment by scdub** [comment]
@anntzer @javidcf Thanks for this fix both here and on Pillow! An application I work on (ArcGIS Pro) also recently "broke the 2<sup>10</sup> module buck" and was causing difficult to assess failures when running from within the application that were not reproducible from standalone Python nor in some cases from within the application. Some digging turned up this issue as the cause, as we were still on releases prior to these fixes being applied. Our base application loads around 400 DLLs, and many Python packages like SciPy have taken the perfectly reasonable route of separating out each component into a separate module, but also means that loading SciPy adds about 100 modules to the running process. Multiplying that across a number of packages can quickly add up to this earlier limit.

**Pull request merged**